src/Raven.Server/Documents/Versioning/VersioningStorage.cs

```diff
@@ -0,0 +1,257 @@
+            configuration.TryGet("MaxRevisions", out maxRevisions);
+            var revisionsCount = IncrementCountOfRevisions(context, key, 1);
+            DeleteOldRevisions(context, table, collectionName, key, maxRevisions, revisionsCount);
+
+            byte* lowerKey;
+            int lowerSize;
+            byte* keyPtr;
+            int keySize;
+            DocumentsStorage.GetLowerKeySliceAndStorageKey(context, key, out lowerKey, out lowerSize, out keyPtr, out keySize);
+
+            var newEtag = ++_lastEtag;
+            var newEtagBigEndian = IPAddress.HostToNetworkOrder(newEtag);
+
+            var tbv = new TableValueBuilder
+            {
+                {lowerKey, lowerSize},
+                {(byte*)&newEtagBigEndian, sizeof(long)},
+                {keyPtr, keySize},
+                {oldDocument.BasePointer, oldDocument.Size}
+            };
+
+            var insert = table.Insert(tbv);
+        }
+
+        private void DeleteOldRevisions(DocumentsOperationContext context, Table table, string collectionName, string key, int? maxRevisions, long revisionsCount)
+        {
+            if (maxRevisions.HasValue == false || maxRevisions.Value == int.MaxValue)
+                return;
+
+            var numberOfRevisionsToDelete = revisionsCount - maxRevisions.Value;
+            if (numberOfRevisionsToDelete <= 0)
+                return;
+
+            var deletedRevisionsCount = table.DeleteForwardFrom(_docsSchema.Indexes["KeyAndEtag"], key, numberOfRevisionsToDelete);
+            IncrementCountOfRevisions(context, collectionName, key, -deletedRevisionsCount);
+        }
+
+        private long IncrementCountOfRevisions(DocumentsOperationContext context, string collectionName, string key, long delta)
+        {
+            var numbers = context.Transaction.InnerTransaction.ReadTree(collectionName + "_VersioningRevisionsCount");
+            return numbers.Increment(key, delta);
+        }
+
+        private void DeleteCountOfRevisions(DocumentsOperationContext context, string collectionName, string key)
+        {
+            var numbers = context.Transaction.InnerTransaction.ReadTree(collectionName + "_VersioningRevisionsCount");
+            numbers.Delete(key);
+        }
+
+        private void DeleteCollectionCountOfRevisions(DocumentsOperationContext context, string collectionName)
+        {
+            context.Transaction.InnerTransaction.DeleteTree(collectionName + "_VersioningRevisionsCount");
+        }
+
+        public void Delete(DocumentsOperationContext context, string collectionName, string key, Document document, bool isSystemDocument)
+        {
+            if (isSystemDocument)
+                return;
+
+            BlittableJsonReaderObject configuration;
+            if (IsVersioningActive(collectionName, false, out configuration) == false)
+                return;
+
+            bool purgeOnDelete;
+            configuration.TryGet("PurgeOnDelete", out purgeOnDelete);
```

Comment:
Probably a better option would be to use nameof() just in case we would do a rename